docs: make courses docs beginner friendly (#1159)
kantord/LibreLingo: docs/courses/README.md

> # LibreLingo Course Documentation

Author: Well, I've added course.md, module.md to act as reference. So in a way it's allowing users to learn about the course system. We also have this on a `docs/courses/` dir. So I guess calling it "Course Documentation" is not too bad.

I've changed it to your suggestion but let me know what you think.

> # LibreLingo Course Documentation

Author: Reverted the change to previous one to see [if this works](https://github.com/kantord/LibreLingo/pull/1159#discussion_r597487518). If this is troublesome, maybe we can change it in future. Resolving this for now.

This document describes how the course system is laid out on LibreLingo and information on how to contribute to the course data.

This document is the first step whether you want to edit an existing course, contribute a translation to an existing course or create a new one from scratch. This is also a good resource to read if you just want to learn about the course system in general.

---

> **NOTE**: If you are not clear about anything on this document feel free to post on [Discussions](https://github.com/kantord/LibreLingo/discussions) so that the community can help you further.

Author: I like the wording of it. Makes it friendlier. But I think there are a couple of things.
- Recently link text is usually kept short. It's a trend or something, but it's common and getting shorter everyday. People link only one word these days. :) One word seems too short, but we can probably make it shorter.
- "Confused?" in all caps may be ok with bold and normal case instead, without a colon (:).
- We can use "Discussions" instead of "Discussion"

Here is my suggestion:
```
    **Confused?**
Ask people in the [Discussions tab on GitHub](https://github.com/kantord/LibreLingo/discussions)!
```

> **NOTE**: If you are not clear about anything on this document feel free to post on [Discussions](https://github.com/kantord/LibreLingo/discussions) so that the community can help you further.

Reviewer: > Recently link text is usually kept short. It's a trend or something, but it's common and getting shorter everyday. People link only one word these days. :) One word seems too short, but we can probably make it shorter.

I haven't heard about this/noticed it. Why is that a trend? What's the value given by following this trend? Do you have some article on it?

> **NOTE**: If you are not clear about anything on this document feel free to post on [Discussions](https://github.com/kantord/LibreLingo/discussions) so that the community can help you further.

Reviewer: > "Confused?" in all caps may be ok with bold and normal case instead, without a colon (:).

Sure, that makes sense

> **NOTE**: If you are not clear about anything on this document feel free to post on [Discussions](https://github.com/kantord/LibreLingo/discussions) so that the community can help you further.

Reviewer: > We can use "Discussions" instead of "Discussion"

Definitely!

> **NOTE**: If you are not clear about anything on this document feel free to post on [Discussions](https://github.com/kantord/LibreLingo/discussions) so that the community can help you further.

Author: > > Recently link text is usually kept short. It's a trend or something, but it's common and getting shorter everyday. People link only one word these days. :) One word seems too short, but we can probably make it shorter.
> 
> I haven't heard about this/noticed it. Why is that a trend? What's the value given by following this trend? Do you have some article on it?

Well, I can't explain it that much. I just saw it around. I guess it's quicker to create and people are lazy. It's hard to find articles on a trend, but I found this [accessibility related article](https://www.mtu.edu/accessibility/training/web/link-text/) which has great examples. And it has:
> Linked text should make sense **out of context**. This is because screen readers often navigation from link to link, skipping the surrounding text. They will read each link out loud, in sequential order.
> ...
> - Keep linked phrases as short as possible, while still being meaningful out of context. This will save screen reader users time, as less will need to be spoken. For example,
> 
>    -  "make a gift" is preferable to "to make a gift online please click here"
>    -  "contact us" is preferable to "you can contact us by"

I have to say people are too lazy and it can make it less accessible which we don't want, at least not for docs. Probably a middle ground would be best. I think we can shorten the link text as much as possible to an extent that it gives context.

Having that said, I didn't know about this earlier. So feel free to point out any other "click here" type links.

> **NOTE**: If you are not clear about anything on this document feel free to post on [Discussions](https://github.com/kantord/LibreLingo/discussions) so that the community can help you further.

Reviewer: Yeah, keeping it short makes a lot of sense.

I'm not sure if it's really too long:

> [Ask people in the Discussion tab on GitHub]

Perhaps we could re-phrase it this way:

> [Ask people on GitHub Discussions]

Or just

> [Ask people] on GitHub discussions

> **NOTE**: If you are not clear about anything on this document feel free to post on [Discussions](https://github.com/kantord/LibreLingo/discussions) so that the community can help you further.

Author: I think the whole line as a link wouldn't look great. And also would be bad for accessibility as the article suggests.

> [Ask people] on GitHub discussions

I think this one is better. So ultimately it becomes something like this:

```suggestion
**Confused?**
[Ask people](https://github.com/kantord/LibreLingo/discussions) on GitHub Discussions.
```

I actually wanted to make this as a notice to the readers that if anything on the page is not clear, they can ask about it on discussions. So help is always there. And this notice is there on every .md page, to help people get to it faster no matter which page they are on.

I wrote things, so I will obviously be biased towards what I wrote. :) But what do you think? As a reader, do you think it communicates the idea? If it is clear, I can make the change.

> **NOTE**: If you are not clear about anything on this document feel free to post on [Discussions](https://github.com/kantord/LibreLingo/discussions) so that the community can help you further.

Reviewer: > [Ask people] on GitHub discussions

I like this. I personally can't decide what's the ideal tradeoff between brevity and clarity, but your reasoning seems pretty logical.

Maybe we can create a separate issue to research this in detail, but for this pull request I'll say this it's perfect!

---

**Table of Contents:**
- [Navigation](#navigation)
- [Basics](#basics)
  - [Terminology](#terminology)
  - [Exploring the courses](#exploring-courses)
  - [Things to do for new contributors](#things-new-contributors)
- [Courses currently open for contributions](#courses-open)
  - [For English speakers](#courses-from-english)
- [Caution](#caution)
  - [Languages that don't use the Latin alphabet](#non-latin-alphabet)

> ## Navigation

Reviewer: ```suggestion
## Getting started
```

> - If you are totally new to the project, read the [Basics](#basics) section below first. It will give you a general understanding about things.

Reviewer: ```suggestion
- Totally new to LibreLingo? Head to the [Basics](#basics) section! It will give you a general understanding about things.
```

> - If you want to edit existing courses or translate them, go to [this page](editing-courses.md).

Reviewer: ```suggestion
- Want to edit or improve existing courses? [Check out this page about editing courses.](editing-courses.md).
```

Author: I've changed it with minor adjustments:
```
- Want to edit or improve existing courses? Check out [this page about editing courses](editing-courses.md).
```

> - If you want to create new courses (typically suited for course designers or advanced users), go to [this page](creating-courses.md).

Reviewer: ```suggestion
- Want to create a new course? [Read out guide on how to create new courses.](creating-courses.md). (Advanced)
```

Author: Here is my suggestion with minor adjustments:
```md
- Want to create a new course? Read out [this guide on how to create new courses](creating-courses.md). (Advanced)
```

## Basics

### Terminology

> If you want to work on courses you need to have an idea on some basic terminology. We have _Courses_, _Modules_ and _Skills_. In a nutshell:

Reviewer: ```suggestion
You'll need to have an idea on basic terminology used in LibreLingo:
```

> - A **Course** is a set of modules. There is typically one course for each language. e.g. `spanish-from-english`, `french-from-english`.

Reviewer: ```suggestion
- A **Course** is a set of Modules. There is typically one course for each language pair. e.g. Spanish for English speakers, Japanese for Italian speakers, etc.
```

> - A **Module** is a set of lessons. They act like groups to keep lessons more organized.

Reviewer: ```suggestion
- A **Module** is a set of Skills.
```

> - A **Skill** is a lesson with a set of questions.

Reviewer: ```suggestion
- A **Skill** is a small unit of course material. In the web app, the user will typically practice a Skill at a time
```

> So the hierarchy would be: `Courses > Modules > Skills`

Reviewer: ```suggestion
These concepts represent a hierarchical structure to LibreLingo course material: `Courses > Modules > Skills`
```

> There are also languages to and from which users will learn.

Reviewer: ```suggestion

```

> - **Destination language** is the language which you want to test for or teach.

Reviewer: ```suggestion
- The **target language** is the language the course is meant to teach
```

> - **Target audience's language** is the language of the users for which the course is made.

Reviewer: ```suggestion
- The **source langauge** is a language that users of the course are assumed to know
```

> The course directory name is usually structured in `<destination language>-from-<target audience's language>`. e.g. `french-from-english` teaches French to English speakers.

Reviewer: Is this part very relevant here? To me it seems it's more relevant when you are creating a new course

Author: I kept it to make it clear how the course directory names are structured. I thought it's important to know which language serves what purpose in the directory name.

Maybe we can move it to somewhere else? Maybe under the next heading "Exploring the courses" after the line "...you will find courses."?

Reviewer: Yes, it might make sense under "exploring courses"

<a id="exploring-courses"></a>
### Exploring the courses

> To have a better understanding on how courses work, you can explore how the course files are organized on this repository.

Reviewer: ```suggestion
To have a better understanding of how courses work, you can explore how the course files are organized on this repository.
```

# Contributing to LibreLingo courses

This document describes how LibreLingo's course system works and how to contribute to course content.

Understanding this document is the first step if you want to:
- edit or improve an existing course
- contribute a new accepted answer
- create a new course from scratch

This is also a good resource if you just want to learn about the course system in general.

---

**NOTE**: If you are not clear about anything on this document feel free to post on [Discussions](https://github.com/kantord/LibreLingo/discussions) so that the community can help you further.

---

**Table of Contents:**
- [Getting started](#getting-started)
- [Basics](#basics)
  - [Terminology](#terminology)
  - [Exploring the courses](#exploring-courses)
  - [Things to do for new contributors](#things-new-contributors)
- [Courses currently open for contributions](#courses-open)
  - [For English speakers](#courses-from-english)
- [Caution](#caution)
  - [Languages that don't use the Latin alphabet](#non-latin-alphabet)

## Getting started

> - Totally new to LibreLingo? Head to the [Basics](#basics) section! It will give you a general understanding about things.

Reviewer: ```suggestion
- Totally new to LibreLingo? Head to the [Basics](#basics) section! It will give you a general understanding of the course system.
```

`of` sounds more natural + better to be specific than use vague words like `things`

- Want to create a new course? Read out [this guide on how to create new courses](creating-courses.md). (Advanced)

## Basics

### Terminology

You'll need to have an idea on basic terminology used in LibreLingo:

- A **Course** is a set of Modules. There is typically one course for each language pair. e.g. Spanish for English speakers, Japanese for Italian speakers, etc.
- A **Module** is a set of Skills.
- A **Skill** is a small unit of course material. In the web app, the user will typically practice a Skill at a time.

These concepts represent a hierarchical structure to LibreLingo course material: `Courses > Modules > Skills`

- The **target language** is the language the course is meant to teach.
- The **source langauge** is a language that users of the course are assumed to know.

The course directory name is usually structured in `<destination language>-from-<target audience's language>`. e.g. `french-from-english` teaches French to English speakers.

<a id="exploring-courses"></a>
### Exploring the courses

To have a better understanding of how courses work, you can explore how the course files are organized on this repository.

> On the `courses` directory of [this repository](https://github.com/kantord/LibreLingo/tree/main/courses) you will find courses. Under each course you'll find modules. And inside each module you'll find skills, which are saved as individual .yaml files.

Reviewer: ```suggestion
In the `courses` directory of [this repository](https://github.com/kantord/LibreLingo/tree/main/courses) you will find courses. Under each course you'll find modules. And inside each module you'll find skills, which are saved as individual .yaml files.
```

> On the `courses` directory of [this repository](https://github.com/kantord/LibreLingo/tree/main/courses) you will find courses. Under each course you'll find modules. And inside each module you'll find skills, which are saved as individual .yaml files.

Reviewer: ```suggestion
You will find courses in the `courses` directory of [LibreLingo's repository](https://github.com/kantord/LibreLingo/tree/main/courses). In each course you'll find modules. And in each module you'll find skills, which are saved as individual .yaml files.
```

Author: OK. So considering the previous changes this becomes:
```suggestion
You will find courses in the `courses` directory of [LibreLingo's repository](https://github.com/kantord/LibreLingo/tree/main/courses). The course directory name is usually structured in `<destination language>-from-<target audience's language>`. e.g. `french-from-english` teaches French to English speakers. In each course you'll find modules. And in each module you'll find skills, which are saved as individual .yaml files.
```

Now let's look into how they look on the frontend, to the user.

> If you go to the [homepage](https://librelingo.app/), courses are listed there. e.g. You can see a "START LEARNING SPANISH" button. (**NOTE:** Depending on the state and completeness of the courses, all of them might not be shown there on the website.) So basically, there are courses for each language. You can think of "courses" as "languages" on your mind if that's more comfortable to you.

Reviewer: ```suggestion
Go to [LibreLingo web app](https://librelingo.app/). Courses are listed there. e.g. You see a "START LEARNING SPANISH" button. (**NOTE:** Depending on the state and completeness of the courses, all of them might not be shown there on the website.)
```

> When you enter a course page, for example the [Spanish course](https://librelingo.app/course/spanish-from-english/), you'll see headings like Basics, Introduction, Activities. These are modules. Each module contains a set of skills. Modules are merely there to serve the purpose of organizing skills. Think of them as groups of units or chapters.

Reviewer: Instead of 

> Modules are merely there to serve the purpose of organizing skills.

Actually I think it's enough to write

> Modules serve the purpose of organizing skills.

Author: Yes. Maybe you are right. But I've written it like this to give the users comfort that they don't need to think about it that much and make it friendlier.

Usually when we talk face to face we tend to use more words to express what we mean, even if those words are not absolutely necessary. I thought using words like this would ease their understanding of it a little bit.

Reviewer: > Usually when we talk face to face we tend to use more words to express what we mean

That's a good point.

How about:

> Modules organize skills into groups, which is their only purpose.

I think this is even more conversational in style, and uses simpler grammar. In the end it's not much shorter than the original phrase, but maybe easier to read

Author: Sounds more formal to me. But maybe it's just me. :)
No problem, I'll apply the change.

> When you enter a course page, for example the Spanish course [here](https://librelingo.app/course/spanish-from-english/), you'll see headings like Basics, Introduction, Activities. These are modules. And under each module heading there are a list of skills under it. Modules are merely there to serve the purpose of organizing skills and basically nothing else. Think of them as folders for your skills.

Reviewer: ```suggestion
When you enter a course page, for example the [Spanish course](https://librelingo.app/course/spanish-from-english/), you'll see headings like Basics, Introduction, Activities. These are modules. Each module contains a set of skills. Modules are merely there to serve the purpose of organizing skills. Think of them as groups of units or chapters.
```

> "Skills" are the actual lessons focused on a particular topic. Each skill is like a set of questions that follow a particular theme. For example, "Animal" skill has questions relating to animals, "Food" skill has questions relating to foods. Each skill is designed to test the skill on a certain topic.

Reviewer: ```suggestion
"Skills" are units of course content focused on a particular topic. Each skill is a set of new knowledge, such as new words or phrases.  Skills can focus on vocabulary and follow a theme: the "Animal" skill has has words and phrases related animals, the "Food" skill is all about food. Skills can also focus on teaching grammar. When it comes to teaching grammar, your main tool is to teach by example.
```

Author: Couple of things:
- '"Animal" skill has has words' > '"Animal" skill has words'
- "phrases related animals" > "phrases related to animals"
- I think "When it comes to teaching grammar, your main tool is to teach by example." is an excellent point but it would be more appropriate on creating courses/skills md files, since this heading is about exploring the course system.

Author: I've made the first 2 changes and added a new title "Tips for creating good skills" in `skill.md` and put your advice under it. Feel free to suggest more tips to be added. I think this will be useful for new course designers.

Inside a skill YAML file there are some keys that you might want to learn of, which is exaplained in detail [here](course.md).

<a id="things-new-contributors"></a>
### Things you can do as a new contributor

If you are new in contributing to course data, these are the things you can do to get a proper understanding of the courses and have a good time with the project:

- Go to [LibreLingo's Spanish course](https://librelingo.app/course/spanish-from-english) and try out some of the skills. Just get an idea on how the questions are structured and the type of answers made available.

> - Clone this repository or [download it](https://github.com/kantord/LibreLingo/archive/main.zip) and study the `courses` directory and how everything's laid out. Be sure to have a good text editor, such as [Notepad++](https://notepad-plus-plus.org/), [Geany](https://geany.org/) in hand to view or edit files.

Reviewer: Do we really want to encourage people to clone the repository? I think it's not really necessary, because on GitHub, you can simply edit files on the GUI and submit merge requests from there :thinking: 

If you are creating a new course from scratch, it's probably good to know how to clone the repo, but just to add contributions to an existing one, it might not be necessary, no?

Author: I did saw something like that. Some questions:

1. Does it support editing multiple files and include it in one commit? Can the commit be done on multiple sessions? (usually translating is a long process and it may take multiple days/sessions)
2. Can the commit be updated later (amend)?

Reviewer: > Does it support editing multiple files and include it in one commit?

No, probably not. [Unless you use Codespaces I guess](https://github.com/features/codespaces).

> Can the commit be done on multiple sessions? (usually translating is a long process and it may take multiple days/sessions)

Yes, I think you can create a draft PR and keep editing the same file in the PR. Actually that way you can also edit multiple files I think

> Can the commit be updated later (amend)?

I don't think so, however you can have multiple commits in an MR I think, and you can squash them when merging

Author: I had some thought.

> Do we really want to encourage people to clone the repository?

Since this is about just exploring, we can skip cloning anyway. We can use a link to browse:

```suggestion
- [Browse this repository](https://github.com/kantord/LibreLingo/) or [download it](https://github.com/kantord/LibreLingo/archive/main.zip) and study the `courses` directory and how everything's laid out. Be sure to have a good text editor, such as [Notepad++](https://notepad-plus-plus.org/), [Geany](https://geany.org/) in hand to view or edit files.
```

I kept the download as a second option here, since it is faster to browse and people can use their favorite text editor to see file contents in an environment they already use everyday. But let me know what you think.

We also have some cloning instructions in `editing-courses.md`. I personally don't use the GitHub GUI options so I'm not aware of its features. If you think it's more intuitive or easier to use, then we might have to make changes on that file for GitHub GUI. The instructions are meant for the beginners (advanced users probably won't need it), so maybe that could be better. Or we can keep existing cli commands in a separate title. I think you know your user-base better than I do, so I leave the decision up to you.

Author: I've made the change in my last suggested change. Please let me know if this should be discussed more.

And I think we've had a rough discussion about including Git cli instructions on a separate file (on a new MR). So we can probably include GitHub GUI options here on a separate MR too. Please resolve this if this is ok for now.

- Try to learn how data is organized within .yaml files and directories. [Course](course.md), [Module](module.md) and [Skill](skill.md) documentation may help you on this.
- Try to learn how YAML works. (see [here](https://en.wikipedia.org/wiki/YAML#Syntax) or [here](https://blog.stackpath.com/yaml/) or [here](https://www.cloudbees.com/blog/yaml-tutorial-everything-you-need-get-started/))

> - Determine some information about your language.

Reviewer: ```suggestion
- Find some information about the language you're working with
```

Although maybe this is only relevant for people who are creating a new course from scratch

Author: This is also applicable for translations because course editors will have to change the language code, name etc. according to their language.

Author: "Find" seems actually applicable for the BCP 47 point. I think "determine" is a better word for other 2 points. Although I think "determine" if fine for all 3 of them. So I suggest:

```suggestion
- Determine some information about the language you're working with.
```

  - Your language name in English. e.g. French, Ancient Greek
  - Language slug (lowercase and no spaces). e.g. french, ancient-greek
  - IETF BCP 47 tag name for your language, which you can get from [here](http://www.iana.org/assignments/language-subtag-registry). Just look for the `Subtag` for your language. e.g. fr, grc
- If you have questions, feel free to ask on [Discussions](https://github.com/kantord/LibreLingo/discussions).

<a id="courses-open"></a>
## Courses currently open for contributions

> <a id="courses-from-english"></a>

Reviewer: I think these should probably not be necessary, right? I mean the `<a />` tags. I think most markdown processors add those automatically

Author: Yes, but the problem is that they're automatic and long.
- If we ever need to edit a title, it will change the existing anchors which would break existing links already shared. Editing all those occurrences would be a chore and sometimes old messages cannot be edited.
- It gives us control how we want our anchors to be, which word to include. For example, "Courses currently open for contributions" would give us "#courses-currently-open-for-contributions" which is long and unnecessary. Instead we can have "#courses-open" which is friendlier and short.

I saw it first on [this manual](https://github.com/void-linux/void-packages/blob/master/Manual.md) and I found that this is useful in reference documentations. One thing however is that when sharing, the links under "Table of contents" should be shared, not the ones present on the shackle icon. Otherwise it would defeat the purpose.

Reviewer: ok, that makes sense I think! 

I don't think we can fully avoid breaking changes, but I can see that it makes sense to at least try
